sockeye/training.py

```diff
@@ -41,9 +41,23 @@
 logger = logging.getLogger(__name__)
 
 
+def get_symbol_fc():
+    x = mx.sym.Variable(name='x')
+    x = mx.sym.flatten(data=x)
+    sym = mx.sym.FullyConnected(data=x, weight=x, no_bias=True, flatten=True, num_hidden=1)
+    return sym
+
+norm_symbol = get_symbol_fc()
+
+# no fp16 support for mx.nd.norm
+def sym_norm(arr):
+    result = norm_symbol.eval(ctx=mx.gpu(0), x=arr.expand_dims(axis=0))[0]
+    return result.reshape((1,))
+
+
 def global_norm(ndarrays: List[mx.nd.NDArray]) -> float:
     # accumulate in a list, as asscalar is blocking and this way we can run the norm calculation in parallel.
-    norms = [mx.nd.square(mx.nd.norm(arr)) for arr in ndarrays if arr is not None]
+    norms = [sym_norm(arr) for arr in ndarrays if arr is not None]
```

Comment:
why use the symbolic API here? Binding an executor every single time seems rather expensive.

Reply:
Symbol gets created once - it is static. But in general, this hack should go away as soon as mxnet will have proper matrix multiplication for float16.

Reply:
True, the symbol only exists once, but you still bind it every single time. What's the advantage over the imparative API here?

Reply:
At the moment, this is the only way to compute `x*x` with float16.

Reply:
This does not work?
```
def imperative_norm(arr):
  arr = mx.nd.flatten(arr)
  return mx.nd.FullyConnected(data=arr, weight=arr, no_bias=True, flatten=True, num_hidden=1)

imperative_norm(mx.nd.array([[2,2]], dtype='float16'))
```